File: scripts/generate-resume-pdf.py
"""Build the downloadable CV from the /resume Svelte page.

Run this script with Python packages lxml and reportlab installed. The PDF
follows the page content so the two do not drift.
"""

from html import escape
from pathlib import Path

from lxml import html
from reportlab.lib import colors
from reportlab.lib.enums import TA_RIGHT
from reportlab.lib.pagesizes import A4
from reportlab.lib.styles import ParagraphStyle
from reportlab.pdfbase import pdfmetrics
from reportlab.pdfbase.ttfonts import TTFont
from reportlab.platypus import (
    HRFlowable,
    KeepTogether,
    ListFlowable,
    ListItem,
    Paragraph,
    SimpleDocTemplate,
    Spacer,
    Table,
    TableStyle,
)


ROOT = Path(__file__).resolve().parents[1]
SOURCE = ROOT / "src" / "routes" / "resume" / "+page.svelte"
OUTPUT = ROOT / "static" / "ben-bergenwall-cv.pdf"


def text(node):
    return " ".join(node.text_content().split())


def first(node, xpath):
    matches = node.xpath(xpath)
    if not matches:
        raise ValueError(f"CV source is missing {xpath}")
    return matches[0]


def paragraph(value, style):
    return Paragraph(escape(value), style)


def article_flowables(article, styles, content_width):
    name = text(first(article, "./div[contains(@class, 'entry-heading')]/heading"))
    date = text(first(article, "./div[contains(@class, 'entry-heading')]/span"))
    heading = Table(
        [[paragraph(name, styles["entry_name"]), paragraph(date, styles["date"])]],
        colWidths=[content_width * 0.64, content_width * 0.36],
        hAlign="LEFT",
    )
    heading.setStyle(
        TableStyle(
            [
                ("VALIGN", (0, 0), (-1, -1), "TOP"),
                ("LEFTPADDING", (0, 0), (-1, -1), 0),
                ("RIGHTPADDING", (0, 0), (-1, -1), 0),
                ("TOPPADDING", (0, 0), (-1, -1), 0),
                ("BOTTOMPADDING", (0, 0), (-1, -1), 0),
            ]
        )
    )
    result = [heading]
    for role in article.xpath("./p[contains(@class, 'entry-role')]"):
        result.append(paragraph(text(role), styles["role"]))
    for description in article.xpath("./p[contains(@class, 'entry-description')]"):
        result.append(paragraph(text(description), styles["body"]))
    for list_node in article.xpath("./ul"):
        result.append(
            ListFlowable(
                [ListItem(paragraph(text(item), styles["body"]), leftIndent=0) for item in list_node.xpath("./li")],
                bulletType="bullet",
                start="circle",
                leftIndent=13,
                bulletFontName="Segoe",
                bulletFontSize=6,
                spaceBefore=2,
            )
        )
    result.append(Spacer(1, 6))
    return result


def main():
    if not SOURCE.exists():
        raise FileNotFoundError(f"Resume page is missing: {SOURCE}")

    font_dir = Path("C:/Windows/Fonts")
    pdfmetrics.registerFont(TTFont("Segoe", str(font_dir / "segoeui.ttf")))
    pdfmetrics.registerFont(TTFont("Segoe-Bold", str(font_dir / "segoeuib.ttf")))
    pdfmetrics.registerFontFamily("Segoe", normal="Segoe", bold="Segoe-Bold")

    ink = colors.HexColor("#1c2733")
    muted = colors.HexColor("#4d5b68")
    rule = colors.HexColor("#cfd7df")
    styles = {
        "name": ParagraphStyle("name", fontName="Segoe-Bold", fontSize=21, leading=25, textColor=ink, spaceAfter=4),
        "lead": ParagraphStyle("lead", fontName="Segoe", fontSize=10.5, leading=15, textColor=ink, spaceAfter=5),
        "link": ParagraphStyle("link", fontName="Segoe", fontSize=9, leading=13, textColor=muted),
        "section": ParagraphStyle("section", fontName="Segoe-Bold", fontSize=10, leading=14, textColor=ink, spaceBefore=4, spaceAfter=6),
        "entry_name": ParagraphStyle("entry-name", fontName="Segoe-Bold", fontSize=9.5, leading=13, textColor=ink),
        "date": ParagraphStyle("date", fontName="Segoe", fontSize=8.3, leading=12, alignment=TA_RIGHT, textColor=muted),
        "role": ParagraphStyle("role", fontName="Segoe", fontSize=8.8, leading=12, textColor=ink, spaceBefore=2, spaceAfter=2),
        "body": ParagraphStyle("body", fontName="Segoe", fontSize=8.8, leading=12, textColor=muted, spaceAfter=2),
        "skill_name": ParagraphStyle("skill-name", fontName="Segoe-Bold", fontSize=8.8, leading=12.5, textColor=ink),
    }

    document = SimpleDocTemplate(
        str(OUTPUT),
        pagesize=A4,
        leftMargin=46,
        rightMargin=46,
        topMargin=34,
        bottomMargin=30,
        title="Ben Bergenwall - CV",
        author="Ben Bergenwall",
    )
    content_width = A4[0] - document.leftMargin - document.rightMargin

    page = html.fromstring(SOURCE.read_text(encoding="utf-8"))
    resume = first(page, "//div[contains(concat(' ', normalize-space(@class), ' '), ' resume ')]")
    intro = first(resume, "./header")
    name = text(first(intro, "./heading"))
    lead = text(first(intro, "./text"))
    github = first(intro, ".//a[starts-with(@href, 'https://github.com/')]")

    story = [
        paragraph(name, styles["name"]),
        paragraph(lead, styles["lead"]),
        Paragraph(
            f'<link href="{escape(github.get("href"), quote=True)}">{escape(github.get("href").removeprefix("https://"))}</link>',
            styles["link"],
        ),
        Spacer(1, 12),
    ]

    for section in resume.xpath("./section"):
        title = text(first(section, "./heading"))
        section_start = [
            HRFlowable(width="100%", thickness=0.6, color=rule, spaceBefore=2, spaceAfter=4),
            paragraph(title, styles["section"]),
        ]
        articles = section.xpath("./div/article")
        if articles:
            story.append(KeepTogether(section_start + article_flowables(articles[0], styles, content_width)))
            for article in articles[1:]:
                story.append(KeepTogether(article_flowables(article, styles, content_width)))
        elif title == "Profile":
            profile = section.xpath("./div/text")
            story.append(KeepTogether(section_start + [paragraph(text(item), styles["body"]) for item in profile] + [Spacer(1, 2)]))
        elif title == "Technical skills":
            skill_cells = []
            for pair in section.xpath("./dl/div"):
                skill_cells.append(
                    [
                        paragraph(text(first(pair, "./dt")), styles["skill_name"]),
                        paragraph(text(first(pair, "./dd")), styles["body"]),
                    ]
                )
            skills = Table([skill_cells], colWidths=[content_width / 3] * 3, hAlign="LEFT")
            skills.setStyle(
                TableStyle(
                    [
                        ("VALIGN", (0, 0), (-1, -1), "TOP"),
                        ("LEFTPADDING", (0, 0), (-1, -1), 0),
                        ("RIGHTPADDING", (0, 0), (-1, -1), 8),
                        ("TOPPADDING", (0, 0), (-1, -1), 0),
                        ("BOTTOMPADDING", (0, 0), (-1, -1), 0),
                    ]
                )
            )
            story.append(KeepTogether(section_start + [skills]))
        else:
            raise ValueError(f"Unexpected resume section: {title}")

    OUTPUT.parent.mkdir(parents=True, exist_ok=True)
    document.build(story)
    print(f"Generated {OUTPUT}")


if __name__ == "__main__":
    main()

File: scripts/generate-resume-pdf.ts
/** Build the downloadable CV from the /resume Svelte page. Run with Bun. */
/* oxlint-disable eslint/no-magic-numbers -- Point sizes and spacing values define the PDF layout. */
import { createWriteStream } from 'node:fs';
import { mkdir, readFile } from 'node:fs/promises';
import { dirname, resolve as resolvePath } from 'node:path';
import { fileURLToPath } from 'node:url';

import PDFKit from 'pdfkit';
import { type AST, parse } from 'svelte/compiler';

type Element = AST.RegularElement | AST.Component;
type Entry = {
  name: string;
  date: string;
  roles: string[];
  descriptions: string[];
  bullets: string[];
};
type Section =
  | { kind: 'profile'; title: string; paragraphs: string[] }
  | { kind: 'entries'; title: string; entries: Entry[] }
  | { kind: 'skills'; title: string; skills: { name: string; detail: string }[] };
type Resume = { name: string; lead: string; github: string; sections: Section[] };

const source = fileURLToPath(new URL('../src/routes/resume/+page.svelte', import.meta.url));
const defaultOutput = fileURLToPath(new URL('../static/ben-bergenwall-cv.pdf', import.meta.url));
const outputFlag = process.argv.indexOf('--output');
if (outputFlag >= 0 && !process.argv[outputFlag + 1]) {
  throw new Error('--output needs a path.');
}
const output = outputFlag >= 0 ? resolvePath(process.argv[outputFlag + 1]) : defaultOutput;

const colors = { ink: '#1c2733', muted: '#4d5b68', rule: '#cfd7df' };
const margins = { bottom: 30, left: 46, right: 46, top: 34 };
const fonts = { bold: 'Helvetica-Bold', regular: 'Helvetica' };

const elements = (parent: Element | AST.Fragment, name: string): Element[] => {
  const fragment = parent.type === 'Fragment' ? parent : parent.fragment;
  return fragment.nodes.filter(
    (node): node is Element =>
      (node.type === 'RegularElement' || node.type === 'Component') && node.name === name,
  );
};

const first = (parent: Element | AST.Fragment, name: string): Element => {
  const found = elements(parent, name).at(0);
  if (!found) {
    throw new Error(`CV source is missing ${name}.`);
  }
  return found;
};

const attribute = (element: Element, name: string): string => {
  const found = element.attributes.find(
    (item): item is AST.Attribute => item.type === 'Attribute' && item.name === name,
  );
  if (!found || !Array.isArray(found.value)) {
    throw new Error(`CV source is missing a static ${name} attribute.`);
  }
  return found.value
    .map((part) => {
      if (part.type !== 'Text') {
        throw new Error(`CV source has a dynamic ${name} attribute.`);
      }
      return part.data;
    })
    .join('');
};

const text = (element: Element): string => {
  const chunks: string[] = [];
  const visit = (fragment: AST.Fragment): void => {
    for (const node of fragment.nodes) {
      if (node.type === 'Text') {
        chunks.push(node.data);
      } else if (node.type === 'RegularElement' || node.type === 'Component') {
        visit(node.fragment);
      } else if (node.type !== 'Comment') {
        throw new Error(`CV source contains dynamic text in ${element.name}.`);
      }
    }
  };
  visit(element.fragment);
  return chunks.join('').replace(/\s+/g, ' ').trim();
};

const parseResume = (svelte: string): Resume => {
  const root = parse(svelte, { modern: true });
  const resume = elements(root.fragment, 'div').find((element) =>
    attribute(element, 'class').split(/\s+/).includes('resume'),
  );
  if (!resume) {
    throw new Error('CV source is missing the resume container.');
  }
  const intro = first(resume, 'header');
  const githubLink = elements(first(intro, 'nav'), 'a').find((link) =>
    attribute(link, 'href').startsWith('https://github.com/'),
  );
  if (!githubLink) {
    throw new Error('CV source is missing the GitHub link.');
  }

  const sections = elements(resume, 'section').map((section): Section => {
    const title = text(first(section, 'Heading'));
    const content = elements(section, 'div').at(0);
    const articles = content ? elements(content, 'article') : [];
    if (articles.length > 0) {
      return {
        entries: articles.map((article) => {
          const heading = first(article, 'div');
          return {
            bullets: elements(article, 'ul').flatMap((list) => elements(list, 'li').map(text)),
            date: text(first(heading, 'span')),
            descriptions: elements(article, 'p')
              .filter((item) => attribute(item, 'class').includes('entry-description'))
              .map(text),
            name: text(first(heading, 'Heading')),
            roles: elements(article, 'p')
              .filter((item) => attribute(item, 'class').includes('entry-role'))
              .map(text),
          };
        }),
        kind: 'entries',
        title,
      };
    }
    if (title === 'Profile') {
      if (!content) {
        throw new Error('CV source is missing the profile content.');
      }
      return { kind: 'profile', paragraphs: elements(content, 'Text').map(text), title };
    }
    if (title === 'Technical skills') {
      const list = first(section, 'dl');
      return {
        kind: 'skills',
        skills: elements(list, 'div').map((item) => ({
          detail: text(first(item, 'dd')),
          name: text(first(item, 'dt')),
        })),
        title,
      };
    }
    throw new Error(`Unexpected resume section: ${title}`);
  });

  return {
    github: attribute(githubLink, 'href'),
    lead: text(first(intro, 'Text')),
    name: text(first(intro, 'Heading')),
    sections,
  };
};

// oxlint-disable-next-line eslint/max-lines-per-function -- Keep the one-page layout sequence together.
const renderResume = (resume: Resume, document: PDFKit.PDFDocument): void => {
  const left = margins.left;
  const width = document.page.width - margins.left - margins.right;
  const bottom = document.page.height - margins.bottom;
  let y = margins.top;

  const height = (value: string, font: string, size: number, textWidth: number): number => {
    document.font(font).fontSize(size);
    return document.heightOfString(value, { lineGap: 1, width: textWidth });
  };
  const ensureSpace = (needed: number): void => {
    if (y + needed > bottom) {
      document.addPage();
      y = margins.top;
    }
  };
  const draw = (
    value: string,
    options: {
      x?: number;
      width?: number;
      font?: string;
      size?: number;
      color?: string;
      after?: number;
    },
  ): void => {
    const x = options.x ?? left;
    const textWidth = options.width ?? width;
    const font = options.font ?? fonts.regular;
    const size = options.size ?? 8.8;
    const textHeight = height(value, font, size, textWidth);
    ensureSpace(textHeight + (options.after ?? 0));
    document
      .font(font)
      .fontSize(size)
      .fillColor(options.color ?? colors.ink);
    document.text(value, x, y, { lineGap: 1, width: textWidth });
    y += textHeight + (options.after ?? 0);
  };
  const sectionTitle = (title: string): void => {
    ensureSpace(28);
    document
      .strokeColor(colors.rule)
      .lineWidth(0.6)
      .moveTo(left, y)
      .lineTo(left + width, y)
      .stroke();
    y += 7;
    draw(title, { after: 5, font: fonts.bold, size: 10 });
  };
  const entry = (item: Entry): void => {
    const nameWidth = width * 0.64;
    const dateWidth = width - nameWidth;
    const rowHeight = Math.max(
      height(item.name, fonts.bold, 9.5, nameWidth),
      height(item.date, fonts.regular, 8.3, dateWidth),
    );
    ensureSpace(rowHeight + 26);
    document.font(fonts.bold).fontSize(9.5).fillColor(colors.ink);
    document.text(item.name, left, y, { lineGap: 1, width: nameWidth });
    document.font(fonts.regular).fontSize(8.3).fillColor(colors.muted);
    document.text(item.date, left + nameWidth, y, {
      align: 'right',
      lineGap: 1,
      width: dateWidth,
    });
    y += rowHeight + 2;
    for (const role of item.roles) {
      draw(role, { after: 2, size: 8.8 });
    }
    for (const description of item.descriptions) {
      draw(description, { after: 2, color: colors.muted, size: 8.8 });
    }
    for (const bullet of item.bullets) {
      const bulletHeight = height(bullet, fonts.regular, 8.8, width - 16);
      ensureSpace(bulletHeight + 2);
      document
        .fillColor(colors.muted)
        .circle(left + 5, y + 5, 1.5)
        .fill();
      draw(bullet, { after: 2, color: colors.muted, width: width - 13, x: left + 13 });
    }
    y += 5;
  };

  draw(resume.name, { after: 4, font: fonts.bold, size: 21 });
  draw(resume.lead, { after: 5, size: 10.5 });
  const githubLabel = resume.github.replace(/^https:\/\//, '');
  const linkHeight = height(githubLabel, fonts.regular, 9, width);
  document.font(fonts.regular).fontSize(9).fillColor(colors.muted);
  document.text(githubLabel, left, y, { link: resume.github, width });
  y += linkHeight + 12;

  for (const section of resume.sections) {
    sectionTitle(section.title);
    if (section.kind === 'profile') {
      for (const paragraph of section.paragraphs) {
        draw(paragraph, { after: 5, color: colors.muted });
      }
      y += 2;
    } else if (section.kind === 'entries') {
      for (const item of section.entries) {
        entry(item);
      }
    } else {
      const columnWidth = width / section.skills.length;
      const heights = section.skills.map((skill) => {
        const cellWidth = columnWidth - 8;
        return (
          height(skill.name, fonts.bold, 8.8, cellWidth) +
          height(skill.detail, fonts.regular, 8.8, cellWidth) +
          3
        );
      });
      ensureSpace(Math.max(...heights));
      let tallest = 0;
      for (const [index, skill] of section.skills.entries()) {
        const x = left + index * columnWidth;
        const cellWidth = columnWidth - 8;
        const nameHeight = height(skill.name, fonts.bold, 8.8, cellWidth);
        const detailHeight = height(skill.detail, fonts.regular, 8.8, cellWidth);
        tallest = Math.max(tallest, nameHeight + detailHeight + 3);
        document.font(fonts.bold).fontSize(8.8).fillColor(colors.ink);
        document.text(skill.name, x, y, { lineGap: 1, width: cellWidth });
        document.font(fonts.regular).fontSize(8.8).fillColor(colors.muted);
        document.text(skill.detail, x, y + nameHeight + 3, { lineGap: 1, width: cellWidth });
      }
      y += tallest;
    }
  }
  if (y > bottom) {
    throw new Error('The CV content overflowed its final page.');
  }
};

const main = async (): Promise<void> => {
  const resume = parseResume(await readFile(source, 'utf8'));
  await mkdir(dirname(output), { recursive: true });
  const document = new PDFKit({
    info: { Author: 'Ben Bergenwall', Title: 'Ben Bergenwall - CV' },
    margins,
    size: 'A4',
  });
  const stream = createWriteStream(output);
  const complete = new Promise<void>((resolve, reject) => {
    stream.on('finish', resolve);
    stream.on('error', reject);
    document.on('error', reject);
  });
  document.pipe(stream);
  renderResume(resume, document);
  document.end();
  await complete;
  process.stdout.write(`Generated ${output}\n`);
};

await main();
